verifier/verifier.go

```diff
@@ -0,0 +1,494 @@
+
+	// verify integrity first. notation will always verify integrity no matter what the signing scheme is
+	envContent, integrityResult := verifyIntegrity(sigBlob, envelopeMediaType, outcome)
+	outcome.EnvelopeContent = envContent
+	outcome.VerificationResults = append(outcome.VerificationResults, integrityResult)
+	if integrityResult.Error != nil {
+		return integrityResult.Error
+	}
+
+	// check if we need to verify using a plugin
+	var pluginCapabilities []proto.Capability
+	verificationPluginName, err := getVerificationPlugin(&outcome.EnvelopeContent.SignerInfo)
+	// use plugin, but getPluginName returns an error
+	if err != nil && err != errExtendedAttributeNotExist {
+		return err
+	}
+	var installedPlugin plugin.Plugin
+	if err == nil {
+		if _, err := getVerificationPluginMinVersion(&outcome.EnvelopeContent.SignerInfo); err != nil && err != errExtendedAttributeNotExist {
+			return notation.ErrorVerificationInconclusive{Msg: fmt.Sprintf("error while getting plugin minimum version, error: %s", err)}
+		}
+		// TODO verify the plugin's version is equal to or greater than `outcome.SignerInfo.SignedAttributes.HeaderVerificationPluginMinVersion`
+		// https://github.com/notaryproject/notation-go/issues/102
+
+		installedPlugin, err = v.pluginManager.Get(ctx, verificationPluginName)
+		if err != nil {
+			return notation.ErrorVerificationInconclusive{Msg: fmt.Sprintf("error while locating the verification plugin %q, make sure the plugin is installed successfully before verifying the signature. error: %s", verificationPluginName, err)}
+		}
+
+		// filter the "verification" capabilities supported by the installed plugin
+		metadata, err := metadata(ctx, installedPlugin, pluginConfig)
+		if err != nil {
+			return err
+		}
+
+		for _, capability := range metadata.Capabilities {
+			if capability == proto.CapabilityRevocationCheckVerifier || capability == proto.CapabilityTrustedIdentityVerifier {
+				pluginCapabilities = append(pluginCapabilities, capability)
+			}
+		}
+
+		if len(pluginCapabilities) == 0 {
+			return notation.ErrorVerificationInconclusive{Msg: fmt.Sprintf("digital signature requires plugin %q with signature verification capabilities (%q and/or %q) installed", verificationPluginName, proto.CapabilityTrustedIdentityVerifier, proto.CapabilityRevocationCheckVerifier)}
+		}
+	}
+
+	// verify x509 trust store based authenticity
+	authenticityResult := verifyAuthenticity(ctx, trustPolicy, outcome)
+	outcome.VerificationResults = append(outcome.VerificationResults, authenticityResult)
+	if isCriticalFailure(authenticityResult) {
+		return authenticityResult.Error
+	}
+
+	// verify x509 trusted identity based authenticity (only if notation needs to perform this verification rather than a plugin)
+	if !proto.CapabilityTrustedIdentityVerifier.In(pluginCapabilities) {
```

Comment:
It is a bit weird to write code like `!element.In(set)`. Usually, we have
```go
if !slice.Contains(pluginCapabilities, proto.CapabilityTrustedIdentityVerifier) {
```
which can be easily translated to natural language

> If plugin is NOT capable of verifying trusted identity, then ...

Reply:
If we do rewrite the form of `!element.In(set)` to `!slice.Contains()`, we can remove `func (c Capability) In(capabilities []Capability) bool` in `plugin/proto/proto.go`.

Reply:
Good idea. Will do.